src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -462,30 +460,32 @@ internal void Flush()
             /// </summary>
             private void RaiseFunctionMetricEvents()
             {
-                if (_runningFunctions.Count == 0)
+                if (_runningFunctions.IsEmpty)
                 {
                     return;
                 }
 
                 // We only need to raise events here for functions that aren't completed.
                 // Events are raised immediately for completed functions elsewhere.
-                FunctionStartedEvent[] runningFunctionsSnapshot = null;
-                lock (_runningFunctionsSyncLock)
+                // Loop through and prune any completed runs
+                foreach (var possiblyRunning in _runningFunctions)
                 {
-                    // effectively we're pruning all the completed invocations here
-                    _runningFunctions = _runningFunctions.Where(p => !p.Completed).ToList();
-
-                    // create a snapshot within the lock so we can enumerate below
-                    runningFunctionsSnapshot = _runningFunctions.ToArray();
+                    if (possiblyRunning.Value.Completed)
+                    {
+                        _runningFunctions.TryRemove(possiblyRunning.Key, out _);
```

Comment:
Does this not throw a `Collection was modified...` exception?

Reply:
Nope! We're safe since it's a concurrent collection now. This safety mechanism is even _likely_ to race as the metrics flush will often overlap a function's own end statement but that race should be safe and not cause any issues. Please sanity check my thinking there!

Reply:
Yup -- you're right! https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getenumerator?view=net-6.0#remarks